collection.go

```diff
@@ -159,26 +159,10 @@ func (cs *CollectionSpec) RewriteConstants(consts map[string]interface{}) error
 // Returns an error if any of the fields can't be found, or
 // if the same map or program is assigned multiple times.
 func (cs *CollectionSpec) Assign(to interface{}) error {
```

Comment:
True, didn't consider that. Conversely, LoadAndAssign now also supports assigning Specs where it didn't before.

Now I understand the context a bit better, wouldn't it make sense to keep `Collection.Assign()` around and deprecate `CollectionSpec.Assign()` instead? That way, the caller can assign Programs and Maps from Collections, and Program(Spec)s and Map(Specs) from CollectionSpecs.

Reply:
Here is why CollectionSpec.Assign exists:
```Go
// from bpf2go
type fooMapSpecs struct
type fooObjects struct

var specs fooMapSpecs
collSpec.Assign(&specs)

specs.FooBar.KeySize = 42

var objs fooObjects
collSpec.LoadAndAssign(&objs) // uses KeySize = 42
```

In my experience, I need access to fooMapSpecs before loading *Map and *Program to do some fudging. Your suggestion to get rid of CollectionSpec.Assign by folding it into CollectionSpec.LoadAndAssign reads kinda funny:

```Go
...
collSpec.LoadAndAssign(&specs, nil)
...
collSpec.LoadAndAssign(&objs, &Options{})
```

It's also awkward that LoadAndAssign needs options. The user has to understand how LoadAndAssign is implemented to figure out whether they should provide options or not.

I'd rather not bring back Collection.Assign since it seems stricly less useful than LoadAndAssign.

Reply:
> I'd rather not bring back Collection.Assign since it seems stricly less useful than LoadAndAssign.

I think it's still useful in the sense that it allows the userspace program to continue with a custom struct instead of a Collection, but now I think about it, there might not be much point in doing so. If you forget to assign any Maps/Progs, those are silently closed on the next GC if the original Collection is no longer referenced. This could be done on purpose, of course, but in that case, LoadAndAssign would be more efficient since it doesn't create unneeded resources. Is this what you meant?

> It's also awkward that LoadAndAssign needs options. The user has to understand how LoadAndAssign is implemented to figure out whether they should provide options or not.

Isn't this already the case? :stuck_out_tongue: In general, I agree. `CollectionSpec.Assign()`'s purpose is clear to me now.

Assuming we're just deprecating `Collection.Assign()` I'll try to come up with a better solution for breaking up `assignValues()`, or at least for providing a way to get access to its internals (map `assignedTo`).

Reply:
Okay, I've restored the `valueOf` structure and (`CollectAnd`)`Assign`'s original behaviour.